server/lorax_server/utils/adapter.py

```diff
@@ -239,3 +257,140 @@ def enum_string_to_adapter_source(adapter_source: str) -> int:
+    for weight_name in model_weights:
+        end_idx = weight_name.rfind(".weight")
+        key = weight_name[:end_idx]
+        for adapter_weight_name in adapter_weights:
+            if key in adapter_weight_name:
+                # example value: 'base_model.model.model.layers.10.self_attn.v_proj.lora_B.weight'
+                # matrix_type gets the second to last element in the module name, i.e. 'lora_B'
+                matrix_type = adapter_weight_name.split(".")[-2]
+                module_mapping[weight_name][matrix_type] = adapter_weight_name
+                processed_adapter_weight_names.add(adapter_weight_name)
+
+    # merge adapter weights into model weights
+    merged_weights = {}
+    for weight_name, adapter_weight_names in tqdm(
+        module_mapping.items(), desc="Merging adapter weights", total=len(module_mapping)
+    ):
+        # TODO: support adapter types beyond LoRA
+        # TODO: put this on GPU if it is available. This should greatly speedup compute_delta_weight
+        lora_A = adapter_weights[adapter_weight_names["lora_A"]]
+        lora_B = adapter_weights[adapter_weight_names["lora_B"]]
+        delta_weight = compute_delta_weight(
+            lora_A, lora_B, adapter_config.fan_in_fan_out, adapter_config.lora_alpha, adapter_config.r
+        )
+
+        # transpose delta weight if necessary
+        # TODO(geoffrey): I believe this is required when using Conv1D layers (gpt2).
+        # We can likely take this out once we've switched to using Linear layers.
+        if (
+            delta_weight.shape != model_weights[weight_name].shape
+            and delta_weight.T.shape == model_weights[weight_name].shape
+        ):
+            delta_weight = delta_weight.T
+        merged_weights[weight_name] = model_weights[weight_name] + delta_weight
+    return merged_weights, processed_adapter_weight_names
+
+
+def create_merged_weight_files(
+    adapter_id: str,
+    model_id: str,
+    model_weight_filenames: List[Path],
+    adapter_source: str = "hub",
+) -> List[Path]:
+    """Creates merged weight files for the given adapter ID and filenames."""
+    api_token = None
+    if adapter_source == PBASE:
+        api_token = api_token or os.environ.get("PREDIBASE_API_TOKEN")
+        adapter_id = map_pbase_model_id_to_s3(adapter_id, api_token)
+        adapter_source = S3
+
+    download_adapter_weights(adapter_id, adapter_source, api_token=api_token)
+
+    if adapter_source == PBASE:
+        adapter_source = S3
+
+    source = get_model_source(adapter_source, adapter_id)
+    adapter_filenames = source.weight_files()
+
+    adapter_path = get_config_path(adapter_id, adapter_source)
+    adapter_config = LoraConfig.from_pretrained(adapter_path)
+    # if adapter_config.base_model_name_or_path != model_id:
+    #     raise ValueError(f"Adapter '{adapter_id}' is not compatible with model '{model_id}'. "
+    #                      f"Use --model-id '{adapter_config.base_model_name_or_path}' instead.")
```

Comment:
Feels like a reasonable check to add back?

Reply:
@tgaddair is there a reason why you had removed this?

Reply:
Ah makes sense